### PR TITLE
Improve CookieCompliance testing

### PR DESCRIPTION
**Pull request by gregw**
+ Improved handling of CookieCompliance.from method
+ Added tests for request and response cookie handling 
+ Use from in jetty.xml

### CODE REVIEW

**Comment by gregw** [comment]
@olamy these fixes may help with the TCK cookie failure. At the very least the `ServerHttpCookieTest` can be used to recreate the TCK failure to investigate deeper.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java
Comment (lorban) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: I think this javadoc should be reverted. Or the link should be given a text other than `...` and other links should be edited too.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java
Comment (lorban) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java]: I think it would be less confusing and safer to introduce a `InvalidCookieException` that extends from `RuntimeException`.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/RFC6265CookieParser.java
Comment (lorban) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/RFC6265CookieParser.java]: Throw a new `InvalidCookieException` that extends from `RuntimeException`?

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/RFC6265CookieParser.java
Comment (lorban) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/RFC6265CookieParser.java]: Throw a new `InvalidCookieException` that extends from `RuntimeException`?

**Review by lorban — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java]: I'm generally not a huge fan of lots of exception types.... I normally ask "do you catch them and handle them differently".  In this case I am.

So I think either I change it so that any exception thrown from getCookies is caught and converted to a BadMessage(400, cause); OR I introduce an `InvalidCookieException` and catch only that.  Any other issues in the cookie parser will result in a 500 rather than a 400.

**Review by lorban — APPROVED** [review]

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CookieParser.java
Comment (joakime) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CookieParser.java]: Why is this a nested private class?
It's exposed via API already.
Can this be a top level class in jetty-http please?

**Review by joakime — CHANGES_REQUESTED** [review]
Just a niggle that the InvalidCookieException is a declared throwable in some APIs, and it a useful exception type to work with, it shouldn't be a nested private class, it should be it's own type.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CookieParser.java
Comment (gregw) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CookieParser.java]: It's not private.  It is public in the CookieParser interface.   We scope many things under interfaces in the API without making them top level classes.

**Comment by gregw** [comment]
@olamy any chance of testing this against the TCK before the merge?

**Review by joakime — APPROVED** [review]
Not a fan of the hidden inner exception that isn't at least static, but lets get this merged.

**Comment by olamy** [comment]
> @olamy any chance of testing this against the TCK before the merge?

I did but this didn't look to fix the TCK test.

**Comment by gregw** [comment]
@joakime it is not hidden.  It's static by default because it is in an interface. If I add the static keyword, intellij tells me that it is not needed.   It is not different to `Handler.Wrapper`, `Handler.Collection`, `Request.Wrapper`, `Content.Sink`, `Content.Source`. etc.  We now have lots of classes and interfaces scoped inside an interface.

**Comment by gregw** [comment]
I'm putting this PR on hold an updating #9402 to include the bulk of these changes, and then merge them back through from 10.

**Comment by gregw** [comment]
@sbordet @lorban This PR was mostly done in #9402 in jetty-10 and has now been merged forward to jetty-12.  What is left are just a few cleanups to use modern switches and to align the other compliance mode impls.

Can you review and if OK please merge.

**Comment by joakime** [comment]
> @joakime it is not hidden. It's static by default because it is in an interface. If I add the static keyword, intellij tells me that it is not needed. It is not different to `Handler.Wrapper`, `Handler.Collection`, `Request.Wrapper`, `Content.Sink`, `Content.Source`. etc. We now have lots of classes and interfaces scoped inside an interface.

Exceptions will propagate through lots of complex logging scenarios.
The non-static inner, means the outer instance is now tied to the instances of that inner.
This could potentially lead to scenarios which GC is impacted.

The other examples don't propagate through other layers like logging.

**Comment by gregw** [comment]
@joakime it is static.  All classes within an interface are static:
Screenshot from 2023-02-22 08-00-40

**Pull request merged**
